core/src/main/java/org/mskcc/oncotree/crosswalk/MSKConceptCache.java

```diff
@@ -67,20 +70,38 @@ public MSKConcept get(String oncoTreeCode) {
     private void resetCache() {
         logger.info("resetCache() -- clearing Crosswalk MSKConcept cache and refilling");
         oncoTreeCodesToMSKConcepts.clear();
-        List<OncoTreeNode> oncoTreeNodes = oncoTreeRepository.getOncoTree(VersionUtil.getDefaultVersion());
-        for (OncoTreeNode node : oncoTreeNodes) {
-            getFromCrosswalkAndSave(node.getCode());
-        } 
+        // versions are ordered in ascending order by release date
+        for (Version version : VersionUtil.getVersions()) {
+            List<OncoTreeNode> oncoTreeNodes = oncoTreeRepository.getOncoTree(version);
+            for (OncoTreeNode node : oncoTreeNodes) {
+                MSKConcept mskConcept = getFromCrosswalkAndSave(node.getCode());
+                // get all codes defined so far for this topbraid uri and save in history
+                if (topBraidURIsToOncotreeCodes.containsKey(node.getURI())) {
+                    // do not add this code to the history, but add any others
+                    HashSet<String> allButThisNode = new HashSet<String>(topBraidURIsToOncotreeCodes.get(node.getURI()));
+                    allButThisNode.remove(node.getCode());
+                    mskConcept.addHistory(allButThisNode);
```

Comment:
Add the set of previous codes into the MSKConcept cache (organized by oncotree code).

So long as an oncotree code is never "resurrected", we should be ok. I think that if a code were present in version 1, renamed in version 2, renamed in version 3, and renamed back to the starting code in version 4, I think we might be in trouble. We can test this in the future.